src/defer.hpp

```diff
@@ -0,0 +1,20 @@
+#pragma once
+
+template <typename Func>
+class Deferrer {
+public:
+    Deferrer(Func func) : m_func(func) {}
+    ~Deferrer() { m_func(); }
+private:
+    Func m_func;
+};
+
+template <typename Func>
+Deferrer<Func> DeferFunction(Func f) {
+    return Deferrer<Func>(f);
+}
+
+#define defer_1(x, y) x##y
+#define defer_2(x, y) defer_1(x, y)
+#define defer_3(x)    defer_2(x, __COUNTER__)
+#define defer(code)   auto defer_3(_defer_) = ::DeferFunction([&](){code;})
```

Comment:
I'd prefer to stick to goto-style error handling, instead of introducing this custom macro magic.

Reply:
Sure, can do.